r55/src/exec.rs

```diff
@@ -133,32 +136,89 @@ pub fn handle_register<EXT, DB: Database>(handler: &mut EvmHandler<'_, EXT, DB>)
     // execute riscv context or old logic.
     let old_handle = handler.execution.execute_frame.clone();
     handler.execution.execute_frame = Arc::new(move |frame, memory, instraction_table, ctx| {
-        let result = if let Some(Some(riscv_context)) = call_stack.borrow_mut().first_mut() {
+        let depth = call_stack.borrow().len() - 1;
+
+        // use last frame as stack is LIFO
+        let mut result = if let Some(Some(riscv_context)) = call_stack.borrow_mut().last_mut() {
+            debug!(
+                "=== [FRAME-{}] Contract: {} ============-",
+                depth,
+                frame.interpreter().contract.target_address,
+            );
             execute_riscv(riscv_context, frame.interpreter_mut(), memory, ctx)?
         } else {
+            debug!("=== [OLD Handler] ==================--");
             old_handle(frame, memory, instraction_table, ctx)?
         };
 
-        // if it is return pop the stack.
+        // if action is return, pop the stack.
         if result.is_return() {
             call_stack.borrow_mut().pop();
+
+            let is_last = call_stack.borrow().is_empty();
+            debug!(
+                "=== [FRAME-{}] Ouput: RETURN {}",
+                depth,
+                if is_last { "(last)" } else { "" }
+            );
+
+            if !is_last {
+                if let Some(Some(parent)) = call_stack.borrow_mut().last_mut() {
+                    if let Some(return_range) = &parent.returned_data_destiny {
+                        if let InterpreterAction::Return { result: res } = &mut result {
+                            // get allocated memory slice
+                            let return_memory = parent
+                                .emu
+                                .cpu
+                                .bus
+                                .get_dram_slice(return_range.clone())
+                                .expect("unable to get memory from return range");
+
+                            debug!("- Return data: {:?}", res.output);
+                            debug!("- Memory range: {:?}", return_range);
+                            debug!("- Memory size: {}", return_memory.len());
+
+                            // write return data to parent's memory
+                            if res.output.len() == return_memory.len() {
+                                return_memory.copy_from_slice(&res.output);
+                            } else {
+                                warn!("Unexpected return data size!");
+                            }
+                        }
+                    }
```

Comment:
can u unpack a bit what's going on here

Reply:
that's what i was trying to explain in the PR description regarding r55 memory.

my understanding is that in revm we have a concept of shared memory, whereas in this case we have DRAM serving the same purpose (hence why we don't need MLOAD and MSTORE).
since each frame (call) uses an independent riscv context with it's own emulator instance, we need to keep track of the expected return data range location in DRAM (`RVEmu.returned_data_destiny`).

when the parent [FRAME-0] executes a `CALL`, we initialize + cache the memory range:
https://github.com/0xrusowsky/r55/blob/7e8eebd29ce061aea541fe33707b0d00c7c0900f/r55/src/exec.rs#L350-#L356

the child [FRAME-1] computes the output of the call, but it is unaware of the expected return data destiny of the parent. Because of that i thought that the only place where i could "move data between frames" was in `fn handle_register()`, where we manipulate the call stack >> in the line where u added the comment is where i have access to both the returned data and the parent context at the same time.

posting a simplified version of the debug logs to help understand the explanation above
```
// emited in `handle_register` (call stack)
=== [FRAME-0] Contract: 0x114c28ea2CDe99e41D2566B1CfAA64a84f564caf ===========

// emited in `execute_riscv` (risc-v emu)
Starting RISC-V execution:  PC: 0x80300000  Return data dst: None
[Syscall::call - 0xf1]
> Return data will be written to: 2150651684..2150651716

// emited in `handle_register` (call stack)
=== [FRAME-1] Contract: 0xf6a171f57ACAc30C292e223eA8aDBb28abD3E14d ===========

// emited in `execute_riscv` (risc-v emu)
Starting RISC-V execution:  PC: 0x80300000  Return data dst: None
[Syscall::return - 0xf3]

// emited in `handle_register` (call stack) >>> this is the step where i copy the output data to the expected location by the parent frame
=== [FRAME-1] Ouput: RETURN
- Return data: 0x000000000000000000000000000000000000000000000000000000000000002a
- Memory range: 2150651684..2150651716
- Memory size: 32

// emited in `handle_register` (call stack)
=== [FRAME-0] Contract: 0x114c28ea2CDe99e41D2566B1CfAA64a84f564caf ===========

// emited in `execute_riscv` (risc-v emu)
Resuming RISC-V execution:  PC: 0x80300994  Return data dst: Some(
    2150651684..2150651716,
)
Loaded return data: 0x000000000000000000000000000000000000000000000000000000000000002

// emited in `handle_register` (call stack)
=== [FRAME-0] Ouput: RETURN (last)
```

Reply:
assuming that what i explained above makes sense... git blame was pointing to Dragan as the editor of most of the `handle_register()` code, so maybe u also want him to validate that my proposed approach is sound

Reply:
based on @leonardoalt comments it looks like my implementation only works with RISCV contracts, as we do need to use shared memory when interacting with EVM contracts